core/sail/api/src/main/java/org/eclipse/rdf4j/sail/helpers/AbstractSailConnection.java

```diff
@@ -375,61 +397,70 @@ public final void prepare() throws SailException {
 
 	@Override
 	public final void commit() throws SailException {
-		if (isActive()) {
-			endUpdate(null);
-		}
-		connectionLock.readLock().lock();
 		try {
-			verifyIsOpen();
+			if (isActive()) {
+				endUpdate(null);
+			}
 
-			updateLock.lock();
+			connectionLock.readLock().lock();
 			try {
-				if (txnActive) {
-					if (!txnPrepared) {
-						prepareInternal();
+				verifyIsOpen();
+
+				updateLock.lock();
+				try {
+					if (txnActive) {
+						if (!txnPrepared) {
+							prepareInternal();
+						}
+						commitInternal();
+						txnActive = false;
+						txnPrepared = false;
 					}
-					commitInternal();
-					txnActive = false;
-					txnPrepared = false;
+				} finally {
+					updateLock.unlock();
 				}
 			} finally {
-				updateLock.unlock();
+				connectionLock.readLock().unlock();
 			}
 		} finally {
-			connectionLock.readLock().unlock();
+			transactionSettings.clear();
 		}
 	}
 
 	@Override
 	public final void rollback() throws SailException {
-		synchronized (added) {
-			added.clear();
-		}
-		synchronized (removed) {
-			removed.clear();
-		}
-		connectionLock.readLock().lock();
 		try {
-			verifyIsOpen();
-
-			updateLock.lock();
+			synchronized (added) {
+				added.clear();
+			}
+			synchronized (removed) {
+				removed.clear();
+			}
+			connectionLock.readLock().lock();
 			try {
-				if (txnActive) {
-					try {
-						rollbackInternal();
-					} finally {
-						txnActive = false;
-						txnPrepared = false;
+				verifyIsOpen();
+
+				updateLock.lock();
+				try {
+					if (txnActive) {
+						try {
+							rollbackInternal();
+						} finally {
+							txnActive = false;
+							txnPrepared = false;
+						}
+					} else {
+						logger.warn("Cannot rollback transaction on connection because transaction is not active",
+								debugEnabled ? new Throwable() : null);
 					}
-				} else {
-					logger.warn("Cannot rollback transaction on connection because transaction is not active",
-							debugEnabled ? new Throwable() : null);
+				} finally {
+					updateLock.unlock();
 				}
 			} finally {
-				updateLock.unlock();
+				connectionLock.readLock().unlock();
 			}
 		} finally {
-			connectionLock.readLock().unlock();
+			transactionSettings.clear();
 		}
 	}
```

Comment:
This is just the git diff engine not understanding that I wrapped everything in try-finally to clear the transaction settings.